File: merge_sort_step.java
import java.io.*;
import java.util.*;

public class merge_sort_step {

    static class Row {
        long number;
        String text;

        Row(long number, String text) {
            this.number = number;
            this.text = text;
        }

        @Override
        public String toString() {
            return number + "/" + text;
        }
    }

    public static void mergeSortWithSteps(List<Row> data, List<String> log) {
        mergeSortInner(data, 0, data.size() - 1, log);
    }

    private static void mergeSortInner(List<Row> arr, int l, int r, List<String> log) {
        if (l < r) {
            int m = (l + r) / 2;
            mergeSortInner(arr, l, m, log);
            mergeSortInner(arr, m + 1, r, log);
            merge(arr, l, m, r, log);
        }
    }

    private static void merge(List<Row> arr, int l, int m, int r, List<String> log) {
        List<Row> left = new ArrayList<>(arr.subList(l, m + 1));
        List<Row> right = new ArrayList<>(arr.subList(m + 1, r + 1));

        int i = 0, j = 0, k = l;

        while (i < left.size() && j < right.size()) {
            if (left.get(i).number < right.get(j).number) {
                arr.set(k++, left.get(i++));
            } else {
                arr.set(k++, right.get(j++));
            }
        }

        while (i < left.size()) arr.set(k++, left.get(i++));
        while (j < right.size()) arr.set(k++, right.get(j++));

        log.add(formatStep(arr));
    }

    private static String formatStep(List<Row> data) {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        for (int i = 0; i < data.size(); i++) {
            sb.append(data.get(i));
            if (i < data.size() - 1) sb.append(", ");
        }
        sb.append("]");
        return sb.toString();
    }

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter CSV filename: ");
        String fileName = sc.nextLine();
        System.out.print("Enter start row: ");
        int startRow = sc.nextInt();
        System.out.print("Enter end row: ");
        int endRow = sc.nextInt();
        sc.close();

        BufferedReader br = new BufferedReader(new FileReader(fileName));
        List<Row> selectedRows = new ArrayList<>();
        String line;
        int currentRow = 0;

        while ((line = br.readLine()) != null) {
            currentRow++;
            if (currentRow >= startRow && currentRow <= endRow) {
                String[] parts = line.split(",");
                selectedRows.add(new Row(Long.parseLong(parts[0]), parts[1]));
            }
        }
        br.close();

        List<String> log = new ArrayList<>();
        
        log.add(formatStep(selectedRows));
        
        mergeSortWithSteps(selectedRows, log);

        String outputFile = String.format("merge_sort_step_%d_%d.txt", startRow, endRow);
        BufferedWriter writer = new BufferedWriter(new FileWriter(outputFile));
        for (String step : log) {
            writer.write(step);
            writer.newLine();
        }
        writer.close();
        System.out.println("Merge steps written to: " + outputFile);
    }
}

File: merge_sort_step.py
import csv


def merge_sort_with_full_steps(data, steps_log, original_size):
    def merge_sort_inner(arr, l, r):
        if l < r:
            m = (l + r) // 2
            merge_sort_inner(arr, l, m)
            merge_sort_inner(arr, m + 1, r)
            merge(arr, l, m, r)

    def merge(arr, l, m, r):
        left = arr[l:m+1]
        right = arr[m+1:r+1]

        i = j = 0
        k = l

        while i < len(left) and j < len(right):
            if left[i][0] < right[j][0]:
                arr[k] = left[i]
                i += 1
            else:
                arr[k] = right[j]
                j += 1
            k += 1

        while i < len(left):
            arr[k] = left[i]
            i += 1
            k += 1

        while j < len(right):
            data[k] = right[j]
            j += 1
            k += 1

        # Log full array after each merge
        step = ', '.join([f"{num}/{txt}" for num, txt in arr])
        steps_log.append(f"[{step}]")

    merge_sort_inner(data, 0, len(data) - 1)


def merge_sort_step(input_file, start_row, end_row):
    output_file = f"merge_sort_step_{start_row}_{end_row}.txt"

    with open(input_file, 'r') as csvfile:
        reader = csv.reader(csvfile)
        rows = [row for idx, row in enumerate(reader) if start_row <= idx + 1 <= end_row]

    data = [(int(row[0]), row[1]) for row in rows]
    steps_log = []

    original = ', '.join([f"{num}/{txt}" for num, txt in data])
    steps_log.append(f"[{original}]")

    merge_sort_with_full_steps(data, steps_log, len(data))

    with open(output_file, 'w') as f:
        for step in steps_log:
            f.write(step + '\n')

    print(f"Merge steps written to: {output_file}")


if __name__ == "__main__":
    file_name = input("Enter CSV filename: ")
    start = int(input("Enter start row: "))
    end = int(input("Enter end row: "))
    merge_sort_step(file_name, start, end)